src/java.base/share/classes/jdk/internal/reflect/ReflectionFactory.java

```diff
@@ -243,6 +243,13 @@ public <T> Constructor<T> copyConstructor(Constructor<T> arg) {
         return langReflectAccess.copyConstructor(arg);
     }
 
+    /**Makes a copy of the passed Class[], unless length is 0 and
+     * the original can safely be reused.
+     */
+    public Class<?>[] copyClasses(Class<?>[] classes) {
```

Comment:
There is no need to make this utility method non-static. This would also obviate the need for having an instance of `ReflectionFactory` to access it. 

Additionally, at the risk of more bikeshedding, I think this utility method is better suited in `AccessibleObject` base class, with package-private visibility. Putting the util methods in `ReflectionFactory` erodes this comment a bit:

```
    <P> The methods in this class are extremely unsafe and can cause
    subversion of both the language and the verifier. For this reason,
    they are all instance methods, and access to the constructor of
    this factory is guarded by a security check, in similar style to
    {@link jdk.internal.misc.Unsafe}. </P>
```

Reply:
Can't this be used by class cloning interfaces too?

Reply:
No, I mean here: https://github.com/openjdk/jdk/blob/4f1a10f84bcfadef263a0890b6834ccd3d5bb52f/src/java.base/share/classes/java/lang/Class.java#L1329

(That's also why I suggest putting the utiltiy method in ReflectionFactory instead of AccessibleObject or Executable)

`Proxy` is meaningless with an empty list of interfaces, as it's solely for implementing interfaces and delegating method calls to the given InvocationHandler.

Reply:
Thanks.  Unfortunately the variable `cloneArray` is actually a method parameter, and most of the callers pass in `false` - so using this utility method to control cloning would actually introduce cloning in several places where it is explicitly avoided.  We may get a performance gain by modifying `Class.getInterfaces()` directly (the sole caller that passes `true`) then eliminating the parameter, then modifying each caller, etc. but that feels like a separate change inspired by this one.

Thoughts?

Reply:
Why can't you do something like this:
```java
return cloneArray ? ReflectionFactory.copyClasses(interfaces) : interfaces;
```

Alternatively, your proposal is a good one too; we can rename this `getInterfaces(boolean)` to `getInterfacesShared()` (like `getSharedEnumConstants()`) and move this copy operation to `getInterfaces()` as you suggest.

Reply:
I really see no reason to try and save on re-use of this one-line method for _everything_. In fact, I do not quite see a very compelling reason to even have the utility method. Sharing the code between `Method` and `Constructor` is clean enough and provides a good balance between reuse and cleanliness. Everything else can have the copy of the method definition, if needed.

Reply:
Alternatively, if a utility method is overkill, we can inline these to `Executable`:
```java
public Class<?>[] getParameterTypes() {
    var shared = getSharedParameterTypes();
    return shared.length == 0 ? shared : shared.clone();
}
```
And the overrides in `Method` and `Constructor` will simply call super; the declarations are kept to preserve the API documentation.

Reply:
I had to read JLS to confirm that changing the `abstract` method to non-abstract one does not break compatibility. 

I am still thinking that we are overthinking this: the readability/maintainability benefits for introducing a one-liner utility method are slim at best. I believe we are spending the disproportionate time on this. So if we cannot agree where to put the utility method -- which implies there is no good place for it -- let's not do it at all. Inline the ternary selector in 4 affected places, and be done with it.